src/rgw/driver/d4n/d4n_policy.cc

```diff
@@ -0,0 +1,445 @@
+  }
+};
+
+template <typename Response, typename CompletionToken>
+auto async_exec(std::shared_ptr<connection> conn,
+                const boost::redis::request& req,
+                Response& resp, CompletionToken&& token)
+{
+  return boost::asio::async_initiate<CompletionToken,
+         void(boost::system::error_code, std::size_t)>(
+      initiate_exec{std::move(conn), req}, token, resp);
+}
+
+template <typename T>
+void redis_exec(std::shared_ptr<connection> conn, boost::system::error_code& ec, boost::redis::request& req, boost::redis::response<T>& resp, optional_yield y)
+{
+  if (y) {
+    auto yield = y.get_yield_context();
+    async_exec(std::move(conn), req, resp, yield[ec]);
+  } else {
+    async_exec(std::move(conn), req, resp, ceph::async::use_blocked[ec]);
+  }
+}
+
+int LFUDAPolicy::set_age(int age, optional_yield y) {
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", "lfuda", "age", std::to_string(age));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec) {
+      return -ec.value();
+    }
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch (std::exception &e) {
+    return -EINVAL;
+  }
+}
+
+int LFUDAPolicy::get_age(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", "lfuda", "age");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec) {
+      return -ec.value();
+    }
+  } catch (std::exception &e) {
+    return -EINVAL;
+  }
+
+  if (!std::get<0>(resp).value()) {
+    if (set_age(1, y)) /* Initialize age */
+      return 1;
+    else
+      return -ENOENT;
+  }
+
+  try { 
+    boost::system::error_code ec;
+    response<std::string> value;
+    request req;
+    req.push("HGET", "lfuda", "age");
+      
+    redis_exec(conn, ec, req, value, y);
+
+    if (ec) {
+      return -ec.value();
+    }
+
+    return std::stoi(std::get<0>(value).value());
+  } catch (std::exception &e) {
+    return -EINVAL;
+  }
+}
+
+int LFUDAPolicy::set_local_weight_sum(size_t weight, optional_yield y) {
+  try {
+    boost::system::error_code ec;
+    response<int> resp;
+    request req;
+    req.push("HSET", dir->cct->_conf->rgw_local_cache_address, "localWeights", std::to_string(weight));
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec) {
+      return -ec.value();
+    }
+
+    return std::get<0>(resp).value(); /* Returns number of fields set */
+  } catch (std::exception &e) {
+    return -EINVAL;
+  }
+}
+
+int LFUDAPolicy::get_local_weight_sum(optional_yield y) {
+  response<int> resp;
+
+  try {
+    boost::system::error_code ec;
+    request req;
+    req.push("HEXISTS", dir->cct->_conf->rgw_local_cache_address, "localWeights");
+
+    redis_exec(conn, ec, req, resp, y);
+
+    if (ec) {
+      return -ec.value();
+    }
+  } catch (std::exception &e) {
+    return -EINVAL;
+  }
+
+  if (!std::get<0>(resp).value()) {
```

Comment:
@samarahu : As part of debugging the hanging `get` call, I discovered that the value of `std::get<0>(resp).value()` is always 1. I believe, the initial value should be 0, and this method will be invoked as part of `update` call, and then the initial local_weight_sum will be set using `sum` calculated below. But that doesn't happen and EINVAL is returned from the next try catch loop, and from this point the local weight sum calculation goes wrong.
Please correct me if I am wrong.

Reply:
I've confirmed that the `set_local_weight_sum` call succeeds even when there is hanging behavior, so there is no EINVAL return here and the method completes successfully, yet the op still hangs. I don't believe this is the source of the bug

Reply:
I am not saying this is the source of the bug, I came across this while debugging the issue. I can share logs with you to take a look at.

Reply:
Note: Run `redis-cli FLUSHALL` to flush redis server of all existing values - this solves the issue.

Reply:
Also as discussed - please modify the argument type of `weight` appropriately in `set_local_weight_sum`